playbooks/install_localhost.yml

```diff
@@ -0,0 +1,14 @@
+---
+# Copyright (c) Ansible Project
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+- name: Install sops on localhost
+  gather_facts: false
+  hosts: localhost
+  tasks:
+    - name: Install sops on localhost
+      ansible.builtin.include_role:
+        name: community.sops.install
+      vars:
+        sops_install_on_localhost: true
```

Comment:
True, this would also work without `sops_install_on_localhost: true`, but then we'd have to change `gather_facts` to `true`, which would gather more facts than we actually need. So using `sops_install_on_localhost: true` together with `gather_facts: false` should be more efficient since only facts that are actually needed are gathered.

Reply:
I think I wasn't completely clear of my intention - I think `sops_install_on_localhost` logic should be removed completely. Right now it is only adding complexity, but no additional functionality (except optionally gathering facts for just localhost) - as far as i can see.

For the fact gathering though, one can use `gather_subset` option along with `gather_facts` to optimize the fact gathering process. Always explicitly gathering these required facts also makes sense, but doing so conditionally, as it is now seems another layer of confusion.

Reply:
Then how do you install sops on the remote system **and** on localhost in the same play, without having this option?

Reply:
Why do you need to do it in the same play? Even if you do need it, you still can:
```yaml
- name: install sops to localhost from any play
  ansible.builtin.import_role:
    name: community.sops.install
  delegate_to: localhost
  run_once: true
```

Reply:
There doesn't seem to be a way to figure out whether a `delegate_to` applies to the current task (and if it does, where it delegates to), or at least nobody in #ansible was able to tell me how to do it. That's why `sops_install_on_localhost` exists.

Reply:
> Why do you need to do it in the same play?

Because ansible's integration tests are a role, so you have to do everything in basically one play.

Reply:
Oh, this revolves around testing again .. and yes, there seems to be no way to detect delegation - at least from vars.

But I still think end-user should use correct targets, opposed to a hack meant for testing.

How would you feel about making this variable private and not instructing users to use this?
Or to move  `ansible.builtin.setup`  to testing code and inject `ansible_facts` from there as a variable, which might be a cleaner approach. I'm only assuming this works.

Also, feel free to resolve this thread if i'm nitpicking too much about stuff i do not understand :)

Reply:
I don't regard this as a 'hack for testing' at all, this is a full-blown feature for me. Installing on the controller is IMO the most important use-case of this role, since in most of the cases you will only need sops there. But if there are more folks who think the `sops_install_on_localhost` option shouldn't be exposed, I will hide it (from the argument spec, so you can still use it, but it's not part of the documented interface).

@gotmax23 @briantist do you have an opinion here? (Or anyone else? :) )

Reply:
I don't have a full opinion because I haven't reviewed it thoroughly yet and I can't right now (maybe later), but I have a few thoughts that also might not be complete or accurate.

I think if the option remains and is usable, there's no reason to hide it from the argspec.

The option seems to serve a specific use that is not (easily) achievable by the end user. I am usually onboard with trying not to devise special cases around things Ansible can already do on its own, and encouraging the use of those methods, but in this case it sounds like encouraging the user to do so requires a lot of explanation and will result in a poor UX, meanwhile the option is straightforward, and does not appear to rely on any hacks or methods that are likely to break in the future.

Reply:
I completely overlooked set_fact usages before, which would set the facts for the remote node either way, unless using localhost as a target (which defeats the purpose of `sops_install_on_localhost`).

Reply:
Indeed, that's a bug. Fixed in e39f18e447e4ea035bb29b1a3cca85cb861b1b3f.

Reply:
There's a few more across other task files in both install and _install_age roles.

Reply:
In the `install` role they only set internal variables, not real facts, so it doesn't really matter where these end up IMO. For `_install_age` I've fixed it in 7d32b311bd39bbb16edeb6361f37d0628c8a0a5e.

Reply:
/reolved by me, unless you are waiting for someone to chip in